tornado/httpclient.py

```diff
@@ -213,6 +213,10 @@ def fetch(self, request, callback=None, raise_error=True, **kwargs):
         If it is a string, we construct an `HTTPRequest` using any additional
         kwargs: ``HTTPRequest(request, **kwargs)``
 
+        This method MUST be called from the ioloop thread, its not
```

Comment:
This may give the impression that `fetch()` is unusual in this respect, when in fact almost nothing in tornado is thread-safe. Our documentation of this fact currently lives [on the tornado.web](http://www.tornadoweb.org/en/stable/web.html#thread-safety-notes) page, which was chosen because that is the most common entry point, although it's not ideal. 

Is there one place we could put this documentation that would be discoverable enough for you, or do you think we need to repeat it (or link to it) in various places in the docs?

Reply:
I'd agree with you on the fact that adding this only to the documentation for this method does make it stand out. And the thread safety notes do make sense - although I personally could not deduce from it that the methods I'm using are thread-unsafe. But in my case the problem was related to AsyncHTTPClient and the IOLoop so instead of reading the whole documentation I attacked the problem by reading documentation for AsyncHTTPClient and related stuff. It might be that I did not read carefully enough, but I could not deduce that .fetch() is thread-unsafe until you enlightened me on the issue.
So, In my opinion It would be really helpful if each thread-unsafe method includes in its docstring something like:
```
This method must only be called from the IOLoop thread, when calling from another thread use 'add_callback'.
```
Or like you're saying instead of repeating we can add links. 👍

Reply:
> So, In my opinion It would be really helpful if each thread-unsafe method

The problem is that that's almost everything in Tornado (everything but `IOLoop.add_callback`), and the thread-unsafe non-blocking concurrency model is kind of the point so it would get tedious (and error-prone) to add this warning (even just a link) to everything. (and even if Tornado added this warning, mixing Tornado with the standard `asyncio` library is increasingly common, and `asyncio` doesn't have a warning like this)

Would you have seen a warning on `IOLoop`? That's the one entry point that everyone has to go through (for now - I expect this will shift to the asyncio event loop), and would make a better place for the warning than `tornado.web`.

Reply:
I see now. It would indeed get tedious if we do it that way. I guess then it would make sense to do two things. First, we could add a note saying that everything in tornado, but 'add_callback' is thread unsafe so every method an Async object - like AsyncHTTPClient - must be called from the IOLoop thread. Secondly, we can add a multi threaded example in AsyncHTTPClient. And perhaps even add it to places suggested by @ploxiln if you think it seems necessary!